### PR TITLE
improve hooks.validate

### PR DESCRIPTION
**Pull request by Uzlopak**
I want to make the Hooks Class capable of handling other events too. To make it possible to add in future more events, the validation function will check if a field is an Array. Also making hooks extend from NullObject, so prototype pollution is not possible.

#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: lib/hooks.js [hunk 1]
```diff
@@ -52,9 +52,11 @@ function Hooks () {
   this.preClose = []
 }
 
+Hooks.property = Object.create(null)
+
 Hooks.prototype.validate = function (hook, fn) {
   if (typeof hook !== 'string') throw new FST_ERR_HOOK_INVALID_TYPE()
-  if (supportedHooks.indexOf(hook) === -1) {
+  if (Array.isArray(this[hook]) === false) {
```
Comment (climba03003) [inline lib/hooks.js:59]: Too relax for the check. I would put the supported hook as the static property of `Hooks` class.
So it own `Hooks.ApplicationHooks`, `Hooks.LifeCycleHooks` and `Hooks.CustomHooks`.

Then, it should be extendible in future by adding custom hook name to the `Hooks.CustomHooks`

**Review by climba03003 — DISMISSED** [review]

**Review by mcollina — APPROVED** [review]
LGTM with @climba03003 considerations

**Comment by Uzlopak** [comment]
I personally disagree. By changing the prototype to NullObject, we only have add and validate and the rest are the arrays of the hooks.

**Comment by Uzlopak** [comment]
I quite dont get, why @climba03003 and @mcollina want to separate the hook names into two or three Objects/Arrays. 

In my PR I basically make Hooks inheriting from null property. So Hooks only has the methods `validate` and `add` as non-Arrays. The rest are the hookFn-Arrays. So checking if a hookName is a attribute of Hooks and is an Array is sufficient. 

@climba03003 wrote that the check is too loose (which is disagree). And that he wants three additional attributes: `ApplicationHooks`, `LifeCycleHooks` and `CustomHooks`.

It makes for me not sense to check three different Objects or Arrays to determine if we have those Hooks already or not, if fastify is just having a method called `addHook`. If we would have methods like `addLifecycleHook` or `addApplicationHook` it would make sense to differentiate them.

File: lib/hooks.js [hunk 2]
```diff
@@ -52,9 +52,11 @@ function Hooks () {
   this.preClose = []
 }
 
+Hooks.property = Object.create(null)
```
Comment (Eomm) [inline lib/hooks.js:55]: Property?
```suggestion
Hooks.prototype = Object.create(null)
```

**Review by Eomm — COMMENTED** [review]
Can't see the pro on splitting the hooks too right now

If we want to add a CustomHooks, the split should stay in that PR, IMHO

For sure a static array would be helpful if we expose it. I listed the hooks in different plugins tho

File: lib/hooks.js (hunk 2 above)
Comment (Uzlopak) [reply lib/hooks.js]: Nothing to see here. Please go on...

**Review by Eomm — APPROVED** [review]

**Comment by mcollina** [comment]
> I quite dont get, why @climba03003 and @mcollina want to separate the hook names into two or three Objects/Arrays.

Can you clarify this? I'm not sure what you refer to.

**Comment by Uzlopak** [comment]
You wrote LGTM with @climba03003  considerations. So I assume you want the changes proposed by @climba03003 ?

**Review by metcoder95 — APPROVED** [review]

**Comment by mcollina** [comment]
My concern of the direction of this work is that it might slow us down without us realizing it. I wrote a good part of the hook system to maximize its performance, which meant duplicating some code. I would prefer to have all custom hooks elsewhere.

**Pull request merged**